File: packages/setpieces-py/src/setpieces/penalties.py
"""Penalty award/conversion summaries."""

import collections
from typing import Iterable

from .events import SetPieceEvent


def penalty_summary(events: Iterable[SetPieceEvent]) -> dict:
    """Awarded/scored/saved/missed/post counts and conversion rate (%)
    across a collection of ``SetPieceEvent``s (only ``kind == "penalty"``
    entries are counted)."""
    subtypes = [e.subtype for e in events if e.kind == "penalty"]
    counts = collections.Counter(subtypes)
    awarded = len(subtypes)
    scored = counts.get("goal", 0)
    return {
        "awarded": awarded,
        "scored": scored,
        "saved": counts.get("saved", 0),
        "missed": counts.get("miss", 0),
        "post": counts.get("post", 0),
        "conversion_rate": (scored / awarded * 100.0) if awarded else None,
    }

File: packages/setpieces-r/setpieces/R/penalties.R
# Penalty award/conversion summaries.

#' Penalty award/conversion summary
#'
#' @param events A list of set-piece event records (from
#'   [extract_set_pieces()]). Only `kind == "penalty"` entries are
#'   counted.
#' @return A named list with `awarded`, `scored`, `saved`, `missed`,
#'   `post`, and `conversion_rate` (percentage, `NA` if none awarded).
#' @export
penalty_summary <- function(events) {
  penalties <- Filter(function(e) identical(e$kind, "penalty"), events)
  subtypes <- vapply(penalties, function(e) e$subtype, character(1))
  awarded <- length(subtypes)
  scored <- sum(subtypes == "goal")

  list(
    awarded = awarded,
    scored = scored,
    saved = sum(subtypes == "saved"),
    missed = sum(subtypes == "miss"),
    post = sum(subtypes == "post"),
    conversion_rate = if (awarded > 0) scored / awarded * 100.0 else NA_real_
  )
}
